src/userdb/userdbctl.c

```diff
@@ -1164,6 +1170,246 @@ static int ssh_authorized_keys(int argc, char *argv[], void *userdata) {
         return r;
 }
 
+static int load_credential_one(int credential_dir_fd, const char *name, int userdb_dir_fd) {
+        int r;
+
+        assert(credential_dir_fd >= 0);
+        assert(name);
+        assert(userdb_dir_fd >= 0);
+
+        const char *user = startswith(name, "userdb.user.");
+        const char *group = startswith(name, "userdb.group.");
+        if (!user && !group)
+                return 0;
+
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        unsigned line = 0, column = 0;
+        r = sd_json_parse_file_at(NULL, credential_dir_fd, name, SD_JSON_PARSE_SENSITIVE, &v, &line, &column);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse credential '%s' as JSON at %u:%u: %m", name, line, column);
+
+        _cleanup_(user_record_unrefp) UserRecord *ur = NULL, *ur_stripped = NULL, *ur_privileged = NULL;
+        _cleanup_(group_record_unrefp) GroupRecord *gr = NULL, *gr_stripped = NULL, *gr_privileged = NULL;
+        _cleanup_free_ char *fn = NULL;
+
+        if (user) {
+                ur = user_record_new();
+                if (!ur)
+                        return log_oom();
+
+                r = user_record_load(ur, v, USER_RECORD_LOAD_MASK_SECRET|USER_RECORD_LOG);
+                if (r < 0)
+                        return r;
+
+                if (!streq(user, ur->user_name))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Credential suffix '%s' does not match user record name '%s'",
+                                               user, ur->user_name);
+
+                if (!uid_is_valid(ur->uid))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "JSON user record missing uid field");
+
+                if (!gid_is_valid(user_record_gid(ur)))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "JSON user record missing gid field");
+
+                r = user_record_clone(ur, USER_RECORD_LOAD_MASK_PRIVILEGED|USER_RECORD_LOG, &ur_stripped);
+                if (r < 0)
+                        return r;
+
+                r = user_record_clone(ur, USER_RECORD_EXTRACT_PRIVILEGED|USER_RECORD_EMPTY_OK|USER_RECORD_LOG, &ur_privileged);
+                if (r < 0)
+                        return r;
+
+                fn = strjoin(ur->user_name, ".user");
+                if (!fn)
+                        return log_oom();
```

Comment:
honestly i would prfer if we check for conflicts here. i.e. just do a quick userdb_by_name() + userdb_by_uid() check here (you can even leave ret as NULL).

do the name check first, and gracefully handle this (i.e. print friendly message, but continue without faulure). do the uid check second, and if there is a conflict fail